pkg/metrics/metrics.go

```diff
@@ -0,0 +1,176 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+
+	pipelinev1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1"
+	"github.com/tektoncd/results/pkg/apis/config"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+	"go.uber.org/zap"
+	"knative.dev/pkg/metrics"
+)
+
+var (
+	registerMutex sync.Mutex
+
+	registeredAt       *time.Time
+	runsNotStoredCount = stats.Int64("runs_not_stored_count", "total number of runs which were deleted without being stored", stats.UnitDimensionless)
+	runsNotStoredView  *view.View
+
+	// Storage latency metric (shared)
+	runStorageLatency     = stats.Float64("run_storage_latency_seconds", "time from run completion to successful storage", stats.UnitSeconds)
+	runStorageLatencyView *view.View
+
+	// Common tags
+	namespaceTag = tag.MustNewKey("namespace")
+	kindTag      = tag.MustNewKey("kind")
+)
+
+// Recorder is used to record metrics for both PipelineRuns and TaskRuns
+type Recorder struct {
+	clock clockwork.Clock
+}
+
+// NewRecorder creates a new metrics recorder instance
+func NewRecorder() *Recorder {
+	return &Recorder{clock: clockwork.NewRealClock()}
+}
+
+func registerViews(logger *zap.SugaredLogger) error {
+	runsNotStoredView = &view.View{
+		Description: runsNotStoredCount.Description(),
+		TagKeys:     []tag.Key{kindTag, namespaceTag},
+		Measure:     runsNotStoredCount,
+		Aggregation: view.Count(),
+	}
+
+	// Storage latency view
+	runStorageLatencyView = &view.View{
+		Description: runStorageLatency.Description(),
+		TagKeys:     []tag.Key{kindTag, namespaceTag},
+		Measure:     runStorageLatency,
+		Aggregation: view.Distribution(0.1, 0.5, 1, 2, 5, 10, 30, 60, 120, 300, 600, 1800),
+	}
+
+	logger.Debug("registering shared metrics views")
+	return view.Register(runsNotStoredView, runStorageLatencyView)
+}
+
+func unregisterViews(logger *zap.SugaredLogger) {
+	logger.Debug("unregistering shared metrics views")
+	var viewsToUnregister []*view.View
+	if runsNotStoredView != nil {
+		viewsToUnregister = append(viewsToUnregister, runsNotStoredView)
+	}
+	if runStorageLatencyView != nil {
+		viewsToUnregister = append(viewsToUnregister, runStorageLatencyView)
+	}
+	view.Unregister(viewsToUnregister...)
+	registeredAt = nil
+}
+
+// IdempotentRegisterViews Ensures all shared views are registered.
+// Does not unregister views if they have already been registered.
```

Comment:
I do not understand this comment. I see we call unregisterViews() here and what it does is `view.Unregister(viewsToUnregister...)`, so it does unregister views. Maybe the `not` here is redundant?

Reply:
Addressed.